File: server/UserData/User/inventory.cs
using Nanina.UserData.ItemData;

namespace Nanina.UserData
{
    public class Inventory 
    {
        public int inventoryIdCounter {get; set;} = 0;
        public Dictionary<int, Equipment> equipment {get; set;} = []; //use inventory id as key
        public Dictionary<short, Item> items {get; set;} = [];          //use item id as key
        public void AddItem(Item item)
        {
            
            if(items.TryGet(item.id, out var oldItem) == false)
            {
                item.inventoryId = inventoryIdCounter;
                inventoryIdCounter++;
                items[item.id] = item;
            }
            else
            {
                items[oldItem.id].count += item.count;
            }
        }

        public void RemoveItem(Item item)
        {
            RemoveItem(item.id, 1);
        }

        public void RemoveItem(short id, int quantity)
        {
            if(items.TryGet(id, out var item) == false)
            {
                Console.Error.WriteLine($"Remove Item got invalid id: {id}");
                return;
            }
            item.count -= quantity;
            if(item.count <= 0)
                items.Remove(item.id);
            
        }

        public void AddEquipment(Equipment obtainedEquipment) 
        {
            
            obtainedEquipment.inventoryId = inventoryIdCounter;
            inventoryIdCounter++;
            equipment[obtainedEquipment.inventoryId] = obtainedEquipment;
        }
    }
}

File: client/src/classes/user/inventory.ts
import type Dictionary from "../dictionary";
import type Equipment from "../item/equipment";
import type Item from "../item/item";


export default class Inventory {
    public equipment!: Dictionary<Equipment>;
    public items!: Dictionary<Item>
    public inventoryIdCounter !: number

    public AddItem(item: Item)
    {
        let oldItem = this.items[item.id]
        if(oldItem == undefined)
        {
            item.inventoryId = this.inventoryIdCounter;
            this.inventoryIdCounter++;
            this.items[item.id] = item;
        }
        else
        {
            this.items[item.id].count += item.count;
        }
    }

    public RemoveItem(item : Item)
    {
        this.RemoveItemWithId(item.id, 1);
    }

    public RemoveItemWithId(id : number, quantity : number)
    {
        let oldItem = this.items[id]
        if(oldItem == undefined)
        {
            console.log("error with removing item with id : ", id)
            return;
        }
        oldItem.count -= quantity;
        if(oldItem.count <= 0)
            delete this.items[id]
        
    }

    public AddEquipment(obtainedEquipment : Equipment) 
    {
        obtainedEquipment.inventoryId = this.inventoryIdCounter;
        this.inventoryIdCounter++;
        this.equipment[obtainedEquipment.inventoryId] = obtainedEquipment;
    }
}







